### ISSUE TITLE
Linking sketches

### ISSUE THREAD

**Issue by thangngoc89** [Sketch-sh/sketch-sh#116]
Idea: https://sketch.sh/s/r7hRFGOXNgZ9riGbmN6ixC/

# Part 1 - JSOO

JSOO provides a pseudo-filesystem that mimics Unix.

- The default working dir is `/static` 
- Cmis files located at `/static/cmis

JSOO provides a module called `Sys_js` for working with the pseudo-filesystem 
http://ocsigen.org/js_of_ocaml/3.1.0/api/Sys_js

What's needed in JSOO side:

- All functions exposed by JSOO located here: https://github.com/Sketch-sh/rtop-evaluator/blob/aac17ef7ca2440d2b7675bacc6b0d62e176ad4ed/evaluator/evaluator.ml#L74

- Expose a new function called createOrUpdate file to allow frontend app to control the file system
  - You would use `Sys_js.read_file` to try to read a file. When it throws `Not_found`, create file, otherwise update file

# Part 2 - Frontend

I'm basically brain dump here so the ideas maybe confusing, feel free the discuss and comment :) 

- There would be a UI for user to paste sketch link (would be nice to have a search in the future) and choose module name for the sketch
- The app will fetch the sketch data and concat all code blocks into a single file.
- Validate if this sketch can compile.
- The data will be stored along with the sketch to reduce loading overhead and concatenating the blocks everytime. 
- Revision timestamp should be stored alongside the code so in the future, we can provide feature like: Hey, there is a new version of your linked sketch, do you want to update it?

- The app will inject the file with `createOrUpdate` method created above. (Check if the file survived a `reset()` call, otherwise we will have to inject the file on every execution)
- Assuming the sketch is linked with under filename `awesome.re`. On each execution, this block will be executed first: 

```
#mod_use "awesome.re"
```

There are many things that need explorations here like:

- Circular dependency: I think we should disallow a linked sketch to be linked with another sketch. Handling dependencies is really sucks.
- Catching error when running `#mod_use` ([it's partly suppressed here](https://github.com/Sketch-sh/sketch-sh/blob/827a128c64cd7f2f58abc312064ec72b79d6ee8f/client/src_analyzer/Worker_Analyze.re#L33))

**Comment by matthiaskern** [comment]
Making good progress with this feature, but would like to get some format the links are stored as.

```reason
type link = {
  revision_at: string,
  note_id: string,
  name: string,
  lang: lang,
  blocks: string
}
```

Also, do we want to make it future-proof for links from outside sketch-sh already?

**Comment by thangngoc89** [comment]
I think it's reasonable to assume that we might allow linking to outside sketch-sh in the future. so let's future proof ourselves with a variant there. 

I don't think we need to put the whole blocks data to link. Just concat everything into a single codeblock.

There is another problem: how to concatenate all code blocks without introducing syntax error?

**Comment by matthiaskern** [comment]
@thangngoc89 In which cases could this lead to syntax errors apart from existing ones?

**Comment by thangngoc89** [comment]
Oh. That's very easy. Try this 
https://sketch.sh/s/wBCOQaBrIf2BzaynFNXSh4/

I think this could be handled as AST level. I'll take a look at this when I'm done with fork feature. For now let's just do string concatenation

**Comment by matthiaskern** [comment]
Closed by #174
